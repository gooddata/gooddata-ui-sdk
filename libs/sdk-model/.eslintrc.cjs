// (C) 2020 GoodData Corporation
module.exports = {
    parser: "@typescript-eslint/parser",
    plugins: ["prettier", "sonarjs", "eslint-plugin-tsdoc", "regexp"],
    extends: [
        "@gooddata",
        "plugin:import-esm/recommended",
        "plugin:sonarjs/recommended",
        "plugin:regexp/recommended",
        "../../.eslintrc.js",
    ],
    parserOptions: { tsconfigRootDir: __dirname },
};
