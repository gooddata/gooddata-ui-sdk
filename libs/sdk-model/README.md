# GoodData.UI domain models

[![npm version](https://img.shields.io/npm/v/@gooddata/sdk-model)](https://www.npmjs.com/@gooddata/sdk-model)&nbsp;
[![npm monthly downloads](https://img.shields.io/npm/dm/@gooddata/sdk-model)](https://npmcharts.com/compare/@gooddata/sdk-model?minimal=true)&nbsp;
![typescript](https://img.shields.io/badge/typescript-first-blue?logo=typescript)

This package is a part of the [GoodData.UI SDK](https://sdk.gooddata.com/gooddata-ui/docs/about_gooddataui.html).
To learn more, check [the source monorepo](https://github.com/gooddata/gooddata-ui-sdk).

This package provides domain models for GoodData.UI.

These domain models are backend-agnostic. This makes them reusable across different [Analytical Backend](https://www.npmjs.com/package/@gooddata/sdk-backend-spi) implementations.
The package includes TypeScript type definitions, factory functions, functions to get or set certain properties
of the objects in an immutable way, and more. These are used in both the `@gooddata/sdk-backend-*` and `@gooddata/sdk-ui* packages`.

## License

(C) 2017-2022 GoodData Corporation

This project is under MIT License. See [LICENSE](https://github.com/gooddata/gooddata-ui-sdk/blob/master/libs/sdk-model/LICENSE).
