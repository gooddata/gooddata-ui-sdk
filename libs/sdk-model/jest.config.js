// (C) 2019 GoodData Corporation
import base from "../../common/config/jest/jest.config.base.mjs";

export default {
  ...base,
};