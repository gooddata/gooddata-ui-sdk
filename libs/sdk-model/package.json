{
    "name": "@gooddata/sdk-model",
    "version": "8.12.3-beta.0",
    "author": "GoodData",
    "description": "GoodData Model definitions used by UI components and Backend SPI and its implementations",
    "repository": {
        "type": "git",
        "url": "https://github.com/gooddata/gooddata-ui-sdk.git",
        "directory": "libs/sdk-model"
    },
    "license": "MIT",
    "main": "dist/index.js",
    "module": "esm/index.js",
    "es2015": "esm/index.js",
    "browser": "dist/index.js",
    "typings": "esm/index.d.ts",
    "sideEffects": false,
    "files": [
        "dist/**/*.js",
        "dist/**/*.json",
        "dist/**/*.d.ts",
        "dist/**/*.map",
        "esm/**/*.js",
        "esm/**/*.json",
        "esm/**/*.d.ts",
        "esm/**/*.map"
    ],
    "config": {
        "eslint": "-c .eslintrc.js --ext ts src/"
    },
    "scripts": {
        "clean": "rm -rf ci dist esm coverage *.log && jest --clearCache",
        "dev": "tsc -p tsconfig.dev.json --watch",
        "build": "concurrently \"npm run build-cjs\" \"npm run build-esm\" && npm run api-extractor",
        "build-cjs": "tsc -p tsconfig.build.json",
        "build-esm": "tsc -p tsconfig.build.esm.json",
        "api-extractor": "mkdir -p api && [ -z \"${CI}\" ] && (api-extractor run -l) || (api-extractor run)",
        "test": "jest --watch",
        "test-once": "jest --maxWorkers=${JEST_MAX_WORKERS:-'45%'}",
        "test-ci": "JEST_JUNIT_OUTPUT=./ci/results/test-results.xml jest --ci --config jest.ci.js",
        "eslint": "eslint $npm_package_config_eslint",
        "eslint-ci": "mkdir -p ./ci/results && eslint -f checkstyle -o ci/results/eslint-results.xml $npm_package_config_eslint",
        "prettier-check": "prettier --check '{src,test}/**/*.{ts,tsx,json,scss,md,yaml,html}'",
        "prettier-write": "prettier --write '{src,test}/**/*.{ts,tsx,json,scss,md,yaml,html}'",
        "dep-cruiser": "depcruise --validate .dependency-cruiser.js --output-type err-long src/",
        "dep-cruiser-ci": "depcruise --validate .dependency-cruiser.js --output-type err-long src/",
        "validate": "npm run dep-cruiser && npm run eslint && npm run prettier-check",
        "validate-ci": "npm run dep-cruiser-ci && npm run eslint-ci && npm run prettier-check"
    },
    "dependencies": {
        "json-stable-stringify": "^1.0.1",
        "lodash": "^4.17.19",
        "spark-md5": "^3.0.0",
        "stringify-object": "^3.3.0",
        "ts-invariant": "^0.7.3",
        "tslib": "^2.0.0"
    },
    "devDependencies": {
        "@gooddata/eslint-config": "^4.1.0",
        "@microsoft/api-documenter": "^7.17.0",
        "@microsoft/api-extractor": "^7.20.0",
        "@types/jest": "^27.0.1",
        "@types/json-stable-stringify": "^1.0.32",
        "@types/lodash": "^4.14.158",
        "@types/spark-md5": "^3.0.1",
        "@types/stringify-object": "^3.2.0",
        "@typescript-eslint/eslint-plugin": "^5.45.0",
        "@typescript-eslint/parser": "^5.45.0",
        "concurrently": "^6.0.2",
        "dependency-cruiser": "^12.12.2",
        "eslint": "^8.25.0",
        "eslint-plugin-header": "^3.0.0",
        "eslint-plugin-import": "^2.22.0",
        "eslint-plugin-jest": "^27.1.1",
        "eslint-plugin-no-only-tests": "^2.4.0",
        "eslint-plugin-prettier": "^4.0.0",
        "eslint-plugin-regexp": "^1.9.0",
        "eslint-plugin-sonarjs": "^0.16.0",
        "eslint-plugin-tsdoc": "^0.2.14",
        "jest": "^27.5.1",
        "jest-junit": "^3.0.0",
        "prettier": "~2.5.0",
        "ts-jest": "^27.0.5",
        "typescript": "4.0.2"
    }
}
