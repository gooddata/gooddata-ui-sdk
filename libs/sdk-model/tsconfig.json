{
    "compilerOptions": {
        "target": "es2017",
        "lib": ["es2017", "dom"],
        "module": "NodeNext",
        "moduleResolution": "nodenext",
        "esModuleInterop": true,
        "importHelpers": true,        
        "sourceMap": true,
        "outDir": "./esm",
        "strict": true,
        "noImplicitAny": true,
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "noImplicitReturns": true,
        "removeComments": false
    },

    "exclude": ["**/node_modules","esm"]
}
