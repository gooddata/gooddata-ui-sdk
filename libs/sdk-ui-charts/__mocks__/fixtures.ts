// (C) 2007-2019 GoodData Corporation
import { LegacyExecutionRecording, legacyRecordedDataView } from "@gooddata/sdk-backend-mockingbird";

import * as rec from "./recordings/playlist.js";
import { DataViewFacade } from "@gooddata/sdk-ui";

function legacyRecordedDataFacade(rec: LegacyExecutionRecording): DataViewFacade {
    return DataViewFacade.for(legacyRecordedDataView(rec));
}

//
// new fixtures
//

export const testWorkspace = "testWorkspace";

//
// Area chart fixtures
//

export const areaChartWith3MetricsAndViewByAttribute = legacyRecordedDataFacade(
    rec.AreaChartWith3MetricsAndViewByAttribute,
);

export const areaChartWithMeasureViewByAndStackBy = legacyRecordedDataFacade(
    rec.AreaChartWithMeasureViewByAndStackBy,
);

//
// Bar chart fixtures
//

export const barChartWith4MetricsAndViewByTwoAttributes = legacyRecordedDataFacade(
    rec.BarChartWith4MetricsAndViewByTwoAttributes,
);
export const barChartWithStackByAndViewByAttributes = legacyRecordedDataFacade(
    rec.BarChartWithStackByAndViewByAttributes,
);
export const barChartWithViewByAttribute = legacyRecordedDataFacade(rec.BarChartWithViewByAttribute);

export const barChartWith3MetricsAndViewByAttribute = legacyRecordedDataFacade(
    rec.BarChartWith3MetricsAndViewByAttribute,
);
export const barChartWith3MetricsAndViewByAttributeFunformat = legacyRecordedDataFacade(
    rec.BarChartWith3MetricsAndViewByAttributeFunformat,
);

export const barChartWith3MetricsAndViewByAttributePercInFormat = legacyRecordedDataFacade(
    rec.BarChartWith3MetricsAndViewByAttributePercInFormat,
);

export const barChartWithPopMeasureAndViewByAttribute = legacyRecordedDataFacade(
    rec.BarChartWithPopMeasureAndViewByAttribute,
);

export const barChartWithPopMeasureAndViewByAttributeX6 = legacyRecordedDataFacade(
    rec.BarChartWithPopMeasureAndViewByAttributeX6,
);

export const barChartWithPreviousPeriodMeasure = legacyRecordedDataFacade(
    rec.BarChartWithPreviousPeriodMeasure,
);
export const barChartWithSingleMeasureAndNoAttributes = legacyRecordedDataFacade(
    rec.BarChartWithSingleMeasureAndNoAttributes,
);

export const barChartWithPreviousPeriodMeasureX6 = legacyRecordedDataFacade(
    rec.BarChartWithPreviousPeriodMeasureX6,
);

//
// Bubble chart
//
export const bubbleChartWith2MetricsAndAttributeNoPrimaries = legacyRecordedDataFacade(
    rec.BubbleChartWith2MetricsAndAttributeNoPrimaries,
);

export const bubbleChartWith1Metric = legacyRecordedDataFacade(rec.BubbleChartWith1Metric);

export const bubbleChartWith3MetricsAndAttribute = legacyRecordedDataFacade(
    rec.BubbleChartWith3MetricsAndAttribute,
);
export const bubbleChartWith3MetricsAndAttributeNullsInData = legacyRecordedDataFacade(
    rec.BubbleChartWith3MetricsAndAttributeNullsInData,
);

//
// Combo chart
//
export const comboWithTwoMeasuresAndViewByAttribute = legacyRecordedDataFacade(
    rec.ComboChartWithTwoMeasuresViewByAttribute,
);

export const comboChartWithTwoMeasuresViewByAttributeNoBuckets = legacyRecordedDataFacade(
    rec.ComboChartWithTwoMeasuresViewByAttributeNoBuckets,
);

export const comboChartWithTwoMeasuresViewByAttributePercformat = legacyRecordedDataFacade(
    rec.ComboChartWithTwoMeasuresViewByAttributePercformat,
);

//
//
//
export const headlineWithOneMeasure = legacyRecordedDataFacade(rec.HeadlineWithOneMeasure);
export const headlineWithOneMeasureWithIdentifier = legacyRecordedDataFacade(
    rec.HeadlineWithOneMeasureWithIdentifier,
);
export const headlineWithTwoMeasures = legacyRecordedDataFacade(rec.HeadlineWithTwoMeasures);
export const headlineWithTwoMeasuresWithIdentifier = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresWithIdentifier,
);
export const headlineWithTwoMeasuresFirstEmpty = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresFirstEmpty,
);
export const headlineWithTwoMeasuresSecondEmpty = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresSecondEmpty,
);
export const headlineWithTwoMeasuresBothEmpty = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresBothEmpty,
);
export const headlineWithTwoMeasuresBothZero = legacyRecordedDataFacade(rec.HeadlineWithTwoMeasuresBothZero);
export const headlineWithTwoMeasuresFirstZero = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresFirstZero,
);
export const headlineWithTwoMeasuresSecondZero = legacyRecordedDataFacade(
    rec.HeadlineWithTwoMeasuresSecondZero,
);
export const headlineWithTwoMeasuresBothSame = legacyRecordedDataFacade(rec.HeadlineWithTwoMeasuresBothSame);

//
// Heatmap recordings
//

export const heatMapWithEmptyCells = legacyRecordedDataFacade(rec.HeatMapWithEmptyCells);
export const heatMapWithMetricRowColumn = legacyRecordedDataFacade(rec.HeatMapWithMetricRowColumn);
//
// Pie chart recordings
//

export const pieChartWithMetricsOnly = legacyRecordedDataFacade(rec.PieChartWithMetricsOnly);
export const pieChartWithMetricsOnlyFundata = legacyRecordedDataFacade(rec.PieChartWithMetricsOnlyFundata);

//
// Scatter plot recording
//

export const scatterPlotWith2MetricsAndAttributeNullsInData = legacyRecordedDataFacade(
    rec.ScatterPlotWith2MetricsAndAttributeNullsInData,
);
export const scatterPlotWith2MetricsAndAttributeWithPrimary = legacyRecordedDataFacade(
    rec.ScatterPlotWith2MetricsAndAttributeWithPrimary,
);

//
// Treemap recordings
//

export const treemapWithMetricAndStackByAttribute = legacyRecordedDataFacade(
    rec.TreemapWithMetricAndStackByAttribute,
);
export const treemapWithMetricAndViewByAttribute = legacyRecordedDataFacade(
    rec.TreemapWithMetricAndViewByAttribute,
);
export const treemapWithTwoMetricsAndStackByAttribute = legacyRecordedDataFacade(
    rec.TreemapWithTwoMetricsAndStackByAttribute,
);
export const treemapWithMetricViewByAndStackByAttribute = legacyRecordedDataFacade(
    rec.TreemapWithMetricViewByAndStackByAttribute,
);
