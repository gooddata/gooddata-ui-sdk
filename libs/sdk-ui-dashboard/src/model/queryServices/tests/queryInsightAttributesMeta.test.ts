// (C) 2021 GoodData Corporation
import { beforeEach, describe, it, expect } from "vitest";
import { DashboardTester, preloadedTesterFactory } from "../../tests/DashboardTester.js";
import { EmptyDashboardIdentifier } from "../../tests/fixtures/Dashboard.fixtures.js";
import { addLayoutSection } from "../../commands/index.js";
import { InsightAttributesMeta, queryInsightAttributesMeta } from "../../queries/index.js";
import { insightRef } from "@gooddata/sdk-model";
import {
    PivotTableWithRowAndColumnAttributes,
    TreemapWithOneMeasureAndViewByDateAndSegmentByDate,
    TreemapWithSingleMeasureAndViewByFilteredToOneElement,
} from "../../tests/fixtures/Insights.fixtures.js";
import { createTestInsightItem, TestSectionHeader } from "../../tests/fixtures/Layout.fixtures.js";

describe("query insight attributes meta", () => {
    let Tester: DashboardTester;

    beforeEach(async () => {
        await preloadedTesterFactory((tester) => {
            Tester = tester;
        }, EmptyDashboardIdentifier);
    });

    it("should return metadata for insight with multiple attributes in buckets", async () => {
        await Tester.dispatchAndWaitFor(
            addLayoutSection(-1, TestSectionHeader, [
                createTestInsightItem(PivotTableWithRowAndColumnAttributes),
            ]),
            "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_ADDED",
        );

        const result: InsightAttributesMeta = await Tester.query(
            queryInsightAttributesMeta(insightRef(PivotTableWithRowAndColumnAttributes)),
        );

        expect(result.displayForms.length).toEqual(4);
        expect(result.attributes.length).toEqual(4);
        expect(result).toMatchSnapshot();
    });

    it("should return metadata for insight with attribute filter", async () => {
        await Tester.dispatchAndWaitFor(
            addLayoutSection(-1, TestSectionHeader, [
                createTestInsightItem(TreemapWithSingleMeasureAndViewByFilteredToOneElement),
            ]),
            "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_ADDED",
        );

        const result: InsightAttributesMeta = await Tester.query(
            queryInsightAttributesMeta(insightRef(TreemapWithSingleMeasureAndViewByFilteredToOneElement)),
        );

        expect(result.displayForms.length).toEqual(1);
        expect(result.attributes.length).toEqual(1);
        expect(result).toMatchSnapshot();
    });

    it("should return metadata for insight with date attributes in buckets", async () => {
        await Tester.dispatchAndWaitFor(
            addLayoutSection(-1, TestSectionHeader, [
                createTestInsightItem(TreemapWithOneMeasureAndViewByDateAndSegmentByDate),
            ]),
            "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_ADDED",
        );

        const result: InsightAttributesMeta = await Tester.query(
            queryInsightAttributesMeta(insightRef(TreemapWithOneMeasureAndViewByDateAndSegmentByDate)),
        );

        expect(result.displayForms.length).toEqual(2);
        expect(result.attributes.length).toEqual(2);
        expect(result).toMatchSnapshot();
    });
});
