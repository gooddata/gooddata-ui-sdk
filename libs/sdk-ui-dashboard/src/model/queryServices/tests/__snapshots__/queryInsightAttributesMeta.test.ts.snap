// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`query insight attributes meta > should return metadata for insight with attribute filter 1`] = `
{
  "attributes": [
    {
      "description": "Product",
      "displayForms": [
        {
          "attribute": {
            "identifier": "attr.f_product.product",
            "type": "attribute",
          },
          "description": "Product",
          "id": "attr.f_product.product",
          "ref": {
            "identifier": "attr.f_product.product",
            "type": "displayForm",
          },
          "title": "Product",
          "type": "displayForm",
          "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/attr.f_product.product",
        },
        {
          "attribute": {
            "identifier": "attr.f_product.product",
            "type": "attribute",
          },
          "description": "",
          "id": "label.f_product.product.name",
          "ref": {
            "identifier": "label.f_product.product.name",
            "type": "displayForm",
          },
          "title": "Product Name",
          "type": "displayForm",
          "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/label.f_product.product.name",
        },
      ],
      "id": "attr.f_product.product",
      "ref": {
        "identifier": "attr.f_product.product",
        "type": "attribute",
      },
      "title": "Product",
      "type": "attribute",
      "uri": "/api/v1/entities/workspaces/referenceworkspace/attributes/attr.f_product.product",
    },
  ],
  "displayForms": [
    {
      "attribute": {
        "identifier": "attr.f_product.product",
        "type": "attribute",
      },
      "description": "",
      "id": "label.f_product.product.name",
      "ref": {
        "identifier": "label.f_product.product.name",
        "type": "displayForm",
      },
      "title": "Product Name",
      "type": "displayForm",
      "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/label.f_product.product.name",
    },
  ],
  "usage": {
    "inAttributes": [
      {
        "identifier": "label.f_product.product.name",
        "type": "displayForm",
      },
    ],
    "inFilters": [
      {
        "identifier": "label.f_product.product.name",
        "type": "displayForm",
      },
    ],
    "inMeasureFilters": [],
  },
}
`;

exports[`query insight attributes meta > should return metadata for insight with date attributes in buckets 1`] = `
{
  "attributes": [
    {
      "description": "Year",
      "displayForms": [
        {
          "attribute": {
            "identifier": "dt_closedate_timestamp.year",
            "type": "attribute",
          },
          "description": "Year",
          "id": "dt_closedate_timestamp.year",
          "ref": {
            "identifier": "dt_closedate_timestamp.year",
            "type": "displayForm",
          },
          "title": "Closed - Year",
          "type": "displayForm",
          "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/dt_closedate_timestamp.year",
        },
      ],
      "id": "dt_closedate_timestamp.year",
      "ref": {
        "identifier": "dt_closedate_timestamp.year",
        "type": "attribute",
      },
      "title": "Closed - Year",
      "type": "attribute",
      "uri": "/api/v1/entities/workspaces/referenceworkspace/attributes/dt_closedate_timestamp.year",
    },
    {
      "description": "Year",
      "displayForms": [
        {
          "attribute": {
            "identifier": "dt_oppcreated_timestamp.year",
            "type": "attribute",
          },
          "description": "Year",
          "id": "dt_oppcreated_timestamp.year",
          "ref": {
            "identifier": "dt_oppcreated_timestamp.year",
            "type": "displayForm",
          },
          "title": "Created - Year",
          "type": "displayForm",
          "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/dt_oppcreated_timestamp.year",
        },
      ],
      "id": "dt_oppcreated_timestamp.year",
      "ref": {
        "identifier": "dt_oppcreated_timestamp.year",
        "type": "attribute",
      },
      "title": "Created - Year",
      "type": "attribute",
      "uri": "/api/v1/entities/workspaces/referenceworkspace/attributes/dt_oppcreated_timestamp.year",
    },
  ],
  "displayForms": [
    {
      "attribute": {
        "identifier": "dt_closedate_timestamp.year",
        "type": "attribute",
      },
      "description": "Year",
      "id": "dt_closedate_timestamp.year",
      "ref": {
        "identifier": "dt_closedate_timestamp.year",
        "type": "displayForm",
      },
      "title": "Closed - Year",
      "type": "displayForm",
      "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/dt_closedate_timestamp.year",
    },
    {
      "attribute": {
        "identifier": "dt_oppcreated_timestamp.year",
        "type": "attribute",
      },
      "description": "Year",
      "id": "dt_oppcreated_timestamp.year",
      "ref": {
        "identifier": "dt_oppcreated_timestamp.year",
        "type": "displayForm",
      },
      "title": "Created - Year",
      "type": "displayForm",
      "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/dt_oppcreated_timestamp.year",
    },
  ],
  "usage": {
    "inAttributes": [
      {
        "identifier": "dt_closedate_timestamp.year",
        "type": "displayForm",
      },
      {
        "identifier": "dt_oppcreated_timestamp.year",
        "type": "displayForm",
      },
    ],
    "inFilters": [],
    "inMeasureFilters": [],
  },
}
`;

exports[`query insight attributes meta > should return metadata for insight with multiple attributes in buckets 1`] = `
{
  "attributes": [
    {
      "description": "Product",
      "displayForms": [
        {
          "attribute": {
            "identifier": "attr.f_product.product",
            "type": "attribute",
          },
          "description": "Product",
          "id": "attr.f_product.product",
          "ref": {
            "identifier": "attr.f_product.product",
            "type": "displayForm",
          },
          "title": "Product",
          "type": "displayForm",
          "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/attr.f_product.product",
        },
        {
          "attribute": {
            "identifier": "attr.f_product.product",
            "type": "attribute",
          },
          "description": "",
          "id": "label.f_product.product.name",
          "ref": {
            "identifier": "label.f_product.product.name",
            "type": "displayForm",
          },
          "title": "Product Name",
          "type": "displayForm",
          "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/label.f_product.product.name",
        },
      ],
      "id": "attr.f_product.product",
      "ref": {
        "identifier": "attr.f_product.product",
        "type": "attribute",
      },
      "title": "Product",
      "type": "attribute",
      "uri": "/api/v1/entities/workspaces/referenceworkspace/attributes/attr.f_product.product",
    },
    {
      "description": "Department",
      "displayForms": [
        {
          "attribute": {
            "identifier": "f_owner.department_id",
            "type": "attribute",
          },
          "description": "Department",
          "id": "f_owner.department_id",
          "ref": {
            "identifier": "f_owner.department_id",
            "type": "displayForm",
          },
          "title": "Department",
          "type": "displayForm",
          "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/f_owner.department_id",
        },
        {
          "attribute": {
            "identifier": "f_owner.department_id",
            "type": "attribute",
          },
          "description": "",
          "displayFormType": "GDC.link",
          "id": "f_owner.department_id.departmenthyperlink",
          "ref": {
            "identifier": "f_owner.department_id.departmenthyperlink",
            "type": "displayForm",
          },
          "title": "Department hyperlink",
          "type": "displayForm",
          "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/f_owner.department_id.departmenthyperlink",
        },
      ],
      "id": "f_owner.department_id",
      "ref": {
        "identifier": "f_owner.department_id",
        "type": "attribute",
      },
      "title": "Department",
      "type": "attribute",
      "uri": "/api/v1/entities/workspaces/referenceworkspace/attributes/f_owner.department_id",
    },
    {
      "description": "Stage Name",
      "displayForms": [
        {
          "attribute": {
            "identifier": "attr.f_stage.stagename",
            "type": "attribute",
          },
          "description": "",
          "id": "label.f_stage.stagename.order",
          "ref": {
            "identifier": "label.f_stage.stagename.order",
            "type": "displayForm",
          },
          "title": "Order",
          "type": "displayForm",
          "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/label.f_stage.stagename.order",
        },
        {
          "attribute": {
            "identifier": "attr.f_stage.stagename",
            "type": "attribute",
          },
          "description": "Stage Name",
          "id": "attr.f_stage.stagename",
          "ref": {
            "identifier": "attr.f_stage.stagename",
            "type": "displayForm",
          },
          "title": "Stage Name",
          "type": "displayForm",
          "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/attr.f_stage.stagename",
        },
        {
          "attribute": {
            "identifier": "attr.f_stage.stagename",
            "type": "attribute",
          },
          "description": "",
          "id": "label.f_stage.stagename.stagename",
          "ref": {
            "identifier": "label.f_stage.stagename.stagename",
            "type": "displayForm",
          },
          "title": "Stage Name",
          "type": "displayForm",
          "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/label.f_stage.stagename.stagename",
        },
      ],
      "id": "attr.f_stage.stagename",
      "ref": {
        "identifier": "attr.f_stage.stagename",
        "type": "attribute",
      },
      "title": "Stage Name",
      "type": "attribute",
      "uri": "/api/v1/entities/workspaces/referenceworkspace/attributes/attr.f_stage.stagename",
    },
    {
      "description": "Region",
      "displayForms": [
        {
          "attribute": {
            "identifier": "f_owner.region_id",
            "type": "attribute",
          },
          "description": "",
          "displayFormType": "GDC.link",
          "id": "f_owner.region_id.regionhyperlink",
          "ref": {
            "identifier": "f_owner.region_id.regionhyperlink",
            "type": "displayForm",
          },
          "title": "Region hyperlink",
          "type": "displayForm",
          "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/f_owner.region_id.regionhyperlink",
        },
        {
          "attribute": {
            "identifier": "f_owner.region_id",
            "type": "attribute",
          },
          "description": "Region",
          "id": "f_owner.region_id",
          "ref": {
            "identifier": "f_owner.region_id",
            "type": "displayForm",
          },
          "title": "Region",
          "type": "displayForm",
          "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/f_owner.region_id",
        },
      ],
      "id": "f_owner.region_id",
      "ref": {
        "identifier": "f_owner.region_id",
        "type": "attribute",
      },
      "title": "Region",
      "type": "attribute",
      "uri": "/api/v1/entities/workspaces/referenceworkspace/attributes/f_owner.region_id",
    },
  ],
  "displayForms": [
    {
      "attribute": {
        "identifier": "attr.f_product.product",
        "type": "attribute",
      },
      "description": "",
      "id": "label.f_product.product.name",
      "ref": {
        "identifier": "label.f_product.product.name",
        "type": "displayForm",
      },
      "title": "Product Name",
      "type": "displayForm",
      "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/label.f_product.product.name",
    },
    {
      "attribute": {
        "identifier": "f_owner.department_id",
        "type": "attribute",
      },
      "description": "Department",
      "id": "f_owner.department_id",
      "ref": {
        "identifier": "f_owner.department_id",
        "type": "displayForm",
      },
      "title": "Department",
      "type": "displayForm",
      "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/f_owner.department_id",
    },
    {
      "attribute": {
        "identifier": "attr.f_stage.stagename",
        "type": "attribute",
      },
      "description": "Stage Name",
      "id": "attr.f_stage.stagename",
      "ref": {
        "identifier": "attr.f_stage.stagename",
        "type": "displayForm",
      },
      "title": "Stage Name",
      "type": "displayForm",
      "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/attr.f_stage.stagename",
    },
    {
      "attribute": {
        "identifier": "f_owner.region_id",
        "type": "attribute",
      },
      "description": "Region",
      "id": "f_owner.region_id",
      "ref": {
        "identifier": "f_owner.region_id",
        "type": "displayForm",
      },
      "title": "Region",
      "type": "displayForm",
      "uri": "/api/v1/entities/workspaces/referenceworkspace/labels/f_owner.region_id",
    },
  ],
  "usage": {
    "inAttributes": [
      {
        "identifier": "label.f_product.product.name",
        "type": "displayForm",
      },
      {
        "identifier": "f_owner.department_id",
        "type": "displayForm",
      },
      {
        "identifier": "attr.f_stage.stagename",
        "type": "displayForm",
      },
      {
        "identifier": "f_owner.region_id",
        "type": "displayForm",
      },
    ],
    "inFilters": [],
    "inMeasureFilters": [],
  },
}
`;
